arch/sim/src/sim/up_hcisocket_host.c

```diff
@@ -0,0 +1,181 @@
+ * under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <sys/types.h>
+#include <sys/uio.h>
+#include <sys/socket.h>
+#include <sys/ioctl.h>
+
+#include <stdbool.h>
+#include <stdio.h>
+#include <string.h>
+#include <unistd.h>
+#include <errno.h>
+
+#include "up_internal.h"
+#include "up_hcisocket_host.h"
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+#define BTPROTO_HCI 1
+
+#define HCI_CHANNEL_RAW   0
+#define HCI_CHANNEL_USER  1
+
+#define HCIDEVDOWN  0x400448ca
+
+/****************************************************************************
+ * Private Types
+ ****************************************************************************/
+
+struct sockaddr_hci
+{
+  sa_family_t     hci_family;
+  unsigned short  hci_dev;
+  unsigned short  hci_channel;
+};
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: bthcisock_host_send
+ *
+ * Description:
+ *   Send a Bluetooth packet out via the host user socket.
+ *
+ * Input Parameters:
+ *   fd: Host Bluetooth socket fd
+ *   pkt_type: Packet type as known to the Linux Bluetooth stack
+ *   data: Pointer to the HCI packet
+ *   len: Length of packet
+ *
+ * Returned Value:
+ *   Zero is returned on success; a negated errno value is returned on any
+ *   failure.
+ *
+ ****************************************************************************/
+
+int bthcisock_host_send(int fd, uint8_t pkt_type, uint8_t *data, size_t len)
+{
+  struct iovec iv[2];
+
+  iv[0].iov_base = &pkt_type;
+  iv[0].iov_len = 1;
+  iv[1].iov_base = data;
+  iv[1].iov_len = len;
+
+  while (writev(fd, iv, 2) < 0)
+    {
+      if (errno == EAGAIN || errno == EINTR)
+        continue;
+      return -1;
+    }
+
+  return 0;
+}
+
+/****************************************************************************
+ * Name: bthcisock_host_read
+ *
+ * Description:
+ *   Read from the Host HCI socket interface.
+ *
+ * Input Parameters:
+ *   fd: Host Bluetooth socket fd
+ *   data: Pointer to store HCI packet
+ *   len: Maximum length of packet
+ *
+ * Returned Value:
+ *   Zero is returned on success; a negated errno value is returned on any
+ *   failure.
+ *
+ ****************************************************************************/
+
+int bthcisock_host_read(int fd, uint8_t *type, void *buf, size_t len)
+{
+  int err;
+  while ((err = recv(fd, buf, len, 0)) < 0 && (errno == EINTR));
```

Comment:
Does it make sense to use readv so the first byte of buf don't contain the type? and bthcisock_host_read is symetric with bthcisock_host_send.

Reply:
That's actually a good idea since we don't need to worry about flags anymore.  Unfortunately I can't quite avoid the temporary buffer because the allocation function requires knowing the type, but the length will be more sane since it wont have the extra pad.

I'll make this change and test in the morning.

Reply:
> That's actually a good idea since we don't need to worry about flags anymore. Unfortunately I can't quite avoid the temporary buffer because the allocation function requires knowing the type, but the length will be more sane since it wont have the extra pad.
> 

bt_buf_alloc doesn't really use type execpt saving into bt_buf_s::type field, so we pass a fixed type and overwrite it after bthcisock_host_read return. Actually, I am thinking to add an enum to bt_buf_type_e(e.g. BT_UNKOWN) to indicate this 
special case.

Reply:
I feel like that violates the contract of the function, but I would feel better if we had the extra enum, so I can make that change along side this.

I really don't like this part of the stack because while we have a max frame in NuttX of around 64 bytes Bluetooth allows up to around 1kB and now a 7 byte command is taking up a whole max frame of data. Anyway that is out of scope for this PR...